iceoryx_posh/source/roudi/iceoryx_roudi_components.cpp

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/roudi/iceoryx_roudi_components.hpp"
+
+namespace iox
+{
+namespace roudi
+{
+IceOryxRouDiMemoryManager* IceOryxRouDiComponents::initRouDiMemoryManager() noexcept
+{
+    // this temporary object will create a roudi mqueue and close it immediatelly
+    // if there was an outdated roudi message queue, it will be cleaned up
+    // if there is an outdated mqueue, the start of the apps will be terminated
+    runtime::MqBase::cleanupOutdatedMessageQueue(roudi::MQ_ROUDI_NAME);
+
+    m_rouDiMemoryManager.createAndAnnounceMemory().or_else([](RouDiMemoryManagerError error) {
+        LogFatal() << "Could not create SharedMemory! Error: " << error;
+        errorHandler(Error::kROUDI_COMPONENTS__SHARED_MEMORY_UNAVAILABLE, nullptr, iox::ErrorLevel::FATAL);
```

Comment:
The program continues here and then we return the pointer to an invalid object? Is this code testable?

I would prefer here to maybe return an `expected<IceOryxRouDiMemoryManager*, ErrorType>` to avoid this issue.

Or we could define the error level fatal as non overloadable and that this always has to terminate.

Reply:
If someone configures the errorHandler to be continuing after a fatal error, I can't really help. **Fatal** is clear in the wording, isn't it? I've implemented the solution you suggested for the sake of a cleaner API. However, on the calling side no further error handling is done, just calling `initRouDiMemoryManager().value()`.

Reply:
In our error handling workshop (and thus in error-handling.md) we defined that a fatal error **always leads to a terminate** in the error handler. It may be preceded with additional code execution, but ultimatetely it has to call a terminate. While this is not enforced by the code right now, when someone reconfigures the error handler to not call terminate he violates this assumption.
This reconfiguration must be possible but should only be used  for testing purpose. A fully configurable error handler is another issue and we need to define a proper user interface for this.

An additional terminate call is therefore redundant and would also be inconsistent to all other places were a fatal error occurs. In theory the error handler should be internal now i.e. not be part of the public API and configurable by the user.

I would postpone this to the point were we refine the error handler and the error handling strategy.

From my point of view it helps to think of returning an expected with an error type as some kind of throw and calling the error handler is analogous to a catch clause which can call a restricted subset of exception handlers (i.e. the error handler with different levels). The decision who handles the error is similarly difficult as for exception (i.e. propagate vs. handle). It must be handled at the latest when it is clear that further operation is not possible, which is the case here.